### PR TITLE
Feature store api

### PR DESCRIPTION
**Pull request by Ben-Epstein**
MVP functional Feature Store API
* Create Feature Set
* Create Feature in feature set
* Deploy Feature Set
* Create Training context
* Get training context
* Run feature elimination
* Feature class definition
* Feature Set definition
* Feature Store Definition
* Training Context definition
* Feature Store Python API docs

All functions (barring one blocked by DB bug) are validated against the database Feature Store metadata.

**Ignore the non .py files, they are auto generated for documentation https://pysplice.readthedocs.io/en/feature_store_api/splicemachine.features.html**

### CODE REVIEW

**Review by splicemaahs — APPROVED** [review]

File: splicemachine/features/constants.py [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+        SELECT fset.feature_set_id, table_name, schema_name, Description, pk_columns, pk_types, deployed FROM {FEATURE_STORE_SCHEMA}.feature_set fset
+        INNER JOIN 
+            (
+                SELECT feature_set_id, STRING_AGG(key_column_name,'|') pk_columns, STRING_AGG(key_column_data_type,'|'
+            ) pk_types 
+        FROM {FEATURE_STORE_SCHEMA}.feature_set_key GROUP BY 1) p 
+        ON fset.feature_set_id=p.feature_set_id 
+        """
+    get_training_contexts = f"""
+    SELECT tc.context_id, tc.Name, tc.Description, CAST(SQL_text AS VARCHAR(1000)) context_sql,
```
Comment (sergioferragut) [inline splicemachine/features/constants.py:76]: We're casting the sql_text to VARCHAR(1000) which is likely to truncate some more complex SQL. Probably want to use the max here (32167). Also, if we are casting it to a limited length when using is, perhaps we should just limit it to 32k characters, but ideally we wouldn't want this limit at all.

File: splicemachine/features/constants.py [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+    WHERE schema_name='{{schema}}' and table_name='{{table}}'
+    """
+
+    feature_set_pk_metadata = f"""
+    INSERT INTO {FEATURE_STORE_SCHEMA}.feature_set_key( feature_set_id, key_column_name, key_column_data_type) 
+    VALUES ({{feature_set_id}}, '{{pk_col_name}}', '{{pk_col_type}}')
+    """
+
+    feature_metadata = f"""
+    INSERT INTO {FEATURE_STORE_SCHEMA}.Feature (feature_set_id, Name, Description, feature_data_type, feature_type, Tags)
```
Comment (sergioferragut) [inline splicemachine/features/constants.py:51]: minor nit "Name" and "Description" are using Title case while all others are lowercase.
This is present in columns throughout the SQL statements in this file.

File: splicemachine/features/constants.py [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+                        WHERE {{where}}
+                        GROUP BY 1,2
+                    )match_keys
+                    WHERE ContextKeyMatchCount = KeyCount 
+              )fl
+          )
+    """
+
+    training_context = f"""
+    INSERT INTO {FEATURE_STORE_SCHEMA}.training_context (Name, Description, SQL_text, ts_column, label_column)
```
Comment (sergioferragut) [inline splicemachine/features/constants.py:120]: In all of the INSERT and UPDATE operations we should be updating last_update_ts and last_update_user_id which are in all tables and will be useful for audit purposes down the line. For the user field, we need to decide if the user domain is the same as the database's and if so, we can probably set it to CURRENT_USER.

**Review by sergioferragut — CHANGES_REQUESTED** [review]
Also in general, how do we resolve the install of the feature store metadata schema? I think we left the DDL in a notebook here: https://github.com/splicemachine/demos/blob/feature_store_webinar2/Feature_Store_Webinar/Feature%20Store%20Schema.ipynb

File: splicemachine/features/constants.py (hunk 1 above)
Comment (Ben-Epstein) [reply splicemachine/features/constants.py:76]: im not sure why we're casting at all.. we probably dont need to, the nSDS can manage a CLOB

**Review by edriggers — APPROVED** [review]

**Pull request merged**